modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php

```diff
@@ -0,0 +1,128 @@
+        // upscale it according to the defined multiplier.
+        $target_with = $image->getWidth() * $this->configuration['multiplier'];
+        $target_height = $image->getHeight() * $this->configuration['multiplier'];
+        $upscale = TRUE;
+      }
+    }
+    if (!$image->scale($target_with, $target_height, $upscale)) {
+      $this->logger->error('Image scale failed using the %toolkit toolkit on %path (%mimetype, %dimensions)', [
+        '%toolkit' => $image->getToolkitId(),
+        '%path' => $image->getSource(),
+        '%mimetype' => $image->getMimeType(),
+        '%dimensions' => $image->getWidth() . 'x' . $image->getHeight(),
+      ]);
+      return FALSE;
+    }
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function transformDimensions(array &$dimensions, $uri) {
+    if ($dimensions['width'] && $dimensions['height']) {
+      // If we are not upscaling the image, check to see if it's smaller
+      // than the defined dimensions.
+      if (!$upscale = $this->configuration['upscale']) {
+        if (
+          (!empty($target_with = $this->configuration['width']) && $this->configuration['width'] > $dimensions['width']) ||
+          (!empty($target_height = $this->configuration['height']) && $this->configuration['height'] > $dimensions['height'])
+        ) {
+          // If the image is smaller than the defined dimensions,
+          // upscale it according to the defined multiplier.
+          $target_with = $dimensions['width'] * $this->configuration['multiplier'];
+          $target_height = $dimensions['height'] * $this->configuration['multiplier'];
+          $upscale = TRUE;
+        }
+      }
+      Image::scaleDimensions($dimensions, $target_with, $target_height, $upscale);
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSummary() {
+    // Since we are extending the image scale effect and not altering it in any
+    // major way, we use the same theme.
+    $summary = [
+      '#theme' => 'image_scale_summary',
+      '#data' => $this->configuration,
+    ];
+    $summary += parent::getSummary();
+
+    return $summary;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultConfiguration() {
+    return parent::defaultConfiguration() + [
+      'multiplier' => 2,
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $form = parent::buildConfigurationForm($form, $form_state);
+    $form['multiplier'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Multiplier'),
+      '#options' => [
+        2 => '2x',
+        3 => '3x',
+      ],
+      '#default_value' => $this->configuration['multiplier'],
+      '#required' => TRUE,
+      '#description' => $this->t("The image will be upscaled according to this multiplier if it is smaller than the dimensions defined in the 'Width' and 'Height' properties."),
```

Comment:
Single quotes.

Also we should mention that it only gets *upscaled* if the `upscale` is not unchecked no?